src/ha_api/mod.rs

```diff
@@ -0,0 +1,130 @@
+use crate::config::YamlConfig;
+use either::*;
+use platform_info::{PlatformInfo, Uname};
+use reqwest::Client;
+use serde::{Deserialize, Serialize};
+use std::error;
+
+type Result<T> = std::result::Result<T, Box<dyn error::Error>>;
+
+const VERSION: &str = env!("CARGO_PKG_VERSION");
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct HaEntityAttribute {
+    pub friendly_name: Option<String>,
+}
+#[derive(Serialize, Deserialize, Debug)]
+pub struct HaEntityState {
+    pub attributes: HaEntityAttribute,
+}
+
+#[derive(Serialize, Deserialize)]
+struct RegisterDeviceRequest {
+    device_id: String,
+    app_id: String,
+    app_name: String,
+    app_version: String,
+    device_name: String,
+    manufacturer: String,
+    model: String,
+    os_name: String,
+    os_version: String,
+    supports_encryption: bool,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct RegisterDeviceResponse {
+    cloud_hook_url: Option<String>,
+    remote_ui_url: Option<String>,
+    secret: Option<String>,
+    webhook_id: String,
+}
+#[derive(Serialize, Deserialize, Debug)]
+pub struct GetAccessTokenRequest {
+    grant_type: String,
+    code: String,
+    client_id: String,
+}
+#[derive(Serialize, Deserialize, Debug)]
+pub struct GetAccessTokenResponse {
+    pub access_token: String,
+    pub expires_in: u32,
+    pub refresh_token: String,
+    pub token_type: String,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct GetAccessTokenError {
+    error: String,
+    error_description: String,
+}
+
+pub async fn get_access_token(
+    config: &YamlConfig,
+    code: String,
+) -> Result<Either<GetAccessTokenError, GetAccessTokenResponse>> {
```

Comment:
I'm not sure you need to use `Either` here. You're returning either an error, or a response, which nicely describes `Result` as-is.